File: napsack.py
n = 6
wv = [(2, 3), (1, 2), (3, 6), (2, 1), (1, 3), (5, 85)]
W = 8

dp = [[0] * 10] * 10  # [[0,0,..,0], [0,0,..,0], ..., [0,0,..,0]]

for i in range(0, n):
    for w in range(0, W):
        if w >= wv[i][0]:
            dp[i+1][w] = max(dp[i][w-wv[i][0]] + wv[i][1], dp[i][w])
        else:
            dp[i+1][w] = dp[i][w]

print(dp[n][w])

File: napsack.cpp
#include <iostream>
#include <climits>
#include <algorithm>
using namespace std;

const int n=6, W=8;
const int w[n] = {2,1,3,2,1,5}, v[n]={3,2,6,1,3,85};

int rec(int i, int cap) {
    if (i == n) {
        return 0;
    } else if (cap < w[i]) {
        return rec(i + 1, cap);
    } else {
        return max(
            rec(i + 1, cap),
            rec(i + 1, cap - w[i]) + v[i]
        );
    }
}

int main() {
    // memset(dp, -1, sizeof(dp));
    cout << rec(0, W) << endl;
}
